components/datetime/src/skeleton.rs

```diff
@@ -0,0 +1,408 @@
+
+//! Skeletons are used for pattern matching. See the [`Skeleton`] struct for more information.
+
+use smallvec::SmallVec;
+use std::{cmp::Ordering, convert::TryFrom, fmt};
+
+use crate::fields::{self, Field, FieldLength, FieldSymbol};
+
+#[cfg(feature = "provider_serde")]
+use serde::{de, ser, Deserialize, Deserializer, Serialize};
+
+#[derive(Debug, PartialEq)]
+struct FieldIndex(usize);
+
+/// A `Skeleton` is used to represent what types of `Field`s are present in a `Pattern`. The
+/// ordering of the `Skeleton`'s `Field`s have no bearing on the ordering of the `Field`s and
+/// `Literal`s in the `Pattern`.
+///
+/// A `Skeleton` is a `Vec<Field>`, but with the invariant that it is sorted according to the canonical
+/// sort order. This order is sorted according to the most significant `Field` to the least significant.
+/// For example, a field with a `Minute` symbol would preceed a field with a `Second` symbol.
+/// This order is documented as the order of fields as presented in the
+/// [UTS 35 Date Field Symbol Table](https://unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table)
+///
+/// The `Field`s are only sorted in the `Skeleton` in order to provide a deterministic
+/// serialization strategy, and to provide a faster `Skeleton` matching operation.
+#[derive(Debug, Eq, PartialEq, Clone)]
+pub struct Skeleton(SmallVec<[fields::Field; 5]>);
+
+impl Skeleton {
+    /// The LiteMap<SkeletonV, PatternV1> structs should be ordered by the `Skeleton` canonical
+    /// order. This helps ensure that the skeleton serialization is sorted deterministically.
+    /// This order is determined by the order of the fields listed in UTS 35, which is ordered
+    /// from most significant, to least significant.
+    ///
+    /// https://unicode.org/reports/tr35/tr35-dates.html#Date_Field_Symbol_Table
+    pub fn compare_canonical_order(&self, other: &Skeleton) -> Ordering {
+        let fields_a = &self.0;
+        let fields_b = &other.0;
+        let max_len = fields_a.len().max(fields_b.len());
+
+        for i in 0..max_len {
+            let maybe_field_a = fields_a.get(i);
+            let maybe_field_b = fields_b.get(i);
+
+            match maybe_field_a {
+                Some(field_a) => match maybe_field_b {
+                    Some(field_b) => {
+                        let ordering = compare_fields(field_a, field_b);
+                        if ordering != Ordering::Equal {
+                            return ordering;
+                        }
+                    }
+                    None => {
+                        // Field A has more fields.
+                        return Ordering::Greater;
+                    }
+                },
+                None => {
+                    // Field B has more fields.
+                    return Ordering::Less;
+                }
+            };
+        }
+
+        Ordering::Equal
+    }
+}
+
+fn compare_fields(field_a: &Field, field_b: &Field) -> Ordering {
```

Comment:
Suggestion: Put this code as `impl Ord for field::Field`.  Once you do this, then I *believe* the Skeleton Ord impl should "just work" via `#[derive(Ord)]`.

Reply:
Oh nice, this simplifies this code quite a bit. I'm starting to like leaning into the trait Impls. It makes things semantic and re-usable. So far, I tend to focus on implementing unique functions to do the job at hand.